### PR TITLE
Implement ERC-20 support

### PR DESCRIPTION
**Pull request by chaotixkilla**

### CODE REVIEW

**Comment by chaotixkilla** [comment]
[ERC-20: Add USDT](https://trello.com/c/snrPYh1G/549-erc-20-add-usdt)

File: lib/plugins/wallet/geth/base.js [hunk 1]
```diff
@@ -15,6 +17,9 @@ const paymentPrefixPath = "m/44'/60'/0'/0'"
 const defaultPrefixPath = "m/44'/60'/1'/0'"
 let lastUsedNonces = {}
 
+// const contract = web3.eth.contract(erc20ABIs.usdt, coins.utils.getErc20Token('USDT').contractAddress)
+// console.log(contract)
```
Comment (josepfo) [inline lib/plugins/wallet/geth/base.js]: remove comments

File: lib/plugins/wallet/geth/base.js [hunk 2]
```diff
@@ -62,25 +67,86 @@ function sendCoins (account, toAddress, cryptoAtoms, cryptoCode) {
 }
 
 function checkCryptoCode (cryptoCode) {
-  if (cryptoCode === 'ETH') return Promise.resolve()
+  if (cryptoCode === 'ETH' || coins.utils.isErc20Token(cryptoCode)) {
+    // console.log(cryptoCode)
```
Comment (josepfo) [inline lib/plugins/wallet/geth/base.js]: same as above

**Review by josepfo — CHANGES_REQUESTED** [review]

File: lib/admin/funding.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ const settingsLoader = require('./settings-loader')
 const configManager = require('./config-manager')
 const wallet = require('../wallet')
 const ticker = require('../ticker')
-const coinUtils = require('../coin-utils')
+const { utils } = require('lamassu-coins')
```
Comment (RafaelTaranto) [inline lib/admin/funding.js]:
```suggestion
const { utils as coinUtils } = require('lamassu-coins')
```

File: lib/plugins/common/kraken.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+const PAIRS = {
```
Comment (RafaelTaranto) [inline lib/plugins/common/kraken.js]: I'm pretty sure this file is not needed anymore

File: lib/new-admin/config/accounts.js [hunk 5]
```diff
@@ -24,7 +24,7 @@ const ALL_ACCOUNTS = [
   { code: 'mock-ticker', display: 'Mock (Caution!)', class: TICKER, cryptos: ALL_CRYPTOS, dev: true },
   { code: 'bitcoind', display: 'bitcoind', class: WALLET, cryptos: [BTC] },
   { code: 'no-layer2', display: 'No Layer 2', class: LAYER_2, cryptos: ALL_CRYPTOS },
-  { code: 'infura', display: 'Infura', class: WALLET, cryptos: [ETH] },
+  { code: 'infura', display: 'Infura', class: WALLET, cryptos: [ETH, USDT] },
   { code: 'geth', display: 'geth (DEPRECATED)', class: WALLET, cryptos: [ETH], deprecated: true },
```
Comment (RafaelTaranto) [inline lib/new-admin/config/accounts.js]: Geth will be un-deprecated after a merge, you can add USDT here as well.

File: lib/plugins/erc20/abi/usdt.json [hunk 6]
```diff
@@ -0,0 +1 @@
...":"address"}],"name":"Deprecate","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"name":"feeBasisPoints","type":"uint256"},{"indexed":false,"name":"maxFee","type":"uint256"}],"name":"Params","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"name":"_blackListedUser","type":"address"},{"indexed":false,"name":"_balance","type":"uint256"}],"name":"DestroyedBlackFunds","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"name":"_user","type":"address"}],"name":"AddedBlackList","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"name":"_user","type":"address"}],"name":"RemovedBlackList","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"owner","type":"address"},{"indexed":true,"name":"spender","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Approval","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"from","type":"address"},{"indexed":true,"name":"to","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Transfer","type":"event"},{"anonymous":false,"inputs":[],"name":"Pause","type":"event"},{"anonymous":false,"inputs":[],"name":"Unpause","type":"event"}]
```
Comment (RafaelTaranto) [inline lib/plugins/erc20/abi/usdt.json]: isn't the ABI a definition of contract? If so I don't think there's anything specific to this ABI that relates to USDT, it should be a generic erc20 thing.

File: lib/plugins/exchange/bitstamp/bitstamp.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+const common = require('../../common/bitstamp')
+const { utils } = require('lamassu-coins')
+
+function buy (account, cryptoAtoms, fiatCode, cryptoCode) {
+  return trade('buy', account, cryptoAtoms, fiatCode, cryptoCode)
+}
+
```
Comment (RafaelTaranto) [inline lib/plugins/exchange/bitstamp/bitstamp.js]: another file that can be removed

File: lib/plugins/exchange/itbit/itbit.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
+const common = require('../../common/itbit')
+const { utils } = require('lamassu-coins')
+
+exports.buy = function (account, cryptoAtoms, fiatCode, cryptoCode) {
```
Comment (RafaelTaranto) [inline lib/plugins/exchange/itbit/itbit.js]: another file to remove

File: lib/plugins/exchange/kraken/kraken.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+// Note: Using DeX3/npm-kraken-api to adjust timeout time
+const Kraken = require('kraken-api')
+const _ = require('lodash/fp')
+
```
Comment (RafaelTaranto) [inline lib/plugins/exchange/kraken/kraken.js]: another one

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: lib/plugins/erc20/abi/usdt.json (hunk 6 above)
Comment (chaotixkilla) [reply lib/plugins/erc20/abi/usdt.json]: I believe you're right. My reasoning for this approach is that each token can select which functions from the erc-20 ABI to use, which leads to, for example USDT's ABI being different than, for example, BNB's ABI. I can find the most generic ERC-20 ABI and simply use that here

File: lib/commission-math.js [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
 const configManager = require('./new-config-manager')
-const coinUtils = require('./coin-utils')
+const { utils: coinUtils } = require('lamassu-coins')
 
 function truncateCrypto (cryptoAtoms, cryptoCode) {
   const DECIMAL_PLACES = 3
   if (cryptoAtoms.eq(0)) return cryptoAtoms
 
-  const scale = coinUtils.getCryptoCurrency(cryptoCode).displayScale
+  const scale = utils.getCryptoCurrency(cryptoCode).displayScale
```
Comment (RafaelTaranto) [inline lib/commission-math.js]: Forgot this one.

File: lib/plugins/wallet/geth/base.js [hunk 11]
```diff
@@ -63,28 +65,88 @@ function sendCoins (account, tx, settings, operatorId) {
...
+        data: contract.transfer.getData(toAddress, hex(toSend))
+      }
+
+      const tx = new Tx(rawTx)
+      const privateKey = wallet.getPrivateKey()
+
+      tx.sign(privateKey)
+
+      return '0x' + tx.serialize().toString('hex')
+    })
```
Comment (RafaelTaranto) [inline lib/plugins/wallet/geth/base.js]: Most of this is duplicated code, can't we split off in a couple of functions?

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: lib/plugins/wallet/geth/base.js (hunk 11 above)
Comment (chaotixkilla) [reply lib/plugins/wallet/geth/base.js]: I think splitting off in multiple functions could get confusing fast, so I think I came with a more elegant solution. Will need some changes if we decide on more token standards, but for now I think it works well

File: lib/plugins/erc20/abi/usdt.json (hunk 6 above)
Comment (RafaelTaranto) [reply lib/plugins/erc20/abi/usdt.json]: Oh, do you have example of the differences. AFAIK there's only a single ABI for erc-20 and if you get the one from the spec should work for every erc-20 compliant token.

**Review by RafaelTaranto — APPROVED** [review]

File: lib/blockchain-utils.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+const options = require('./options')
+
+function blockchainDir () {
+  return options.blockchainDir
+}
+
+module.exports = { blockchainDir }
```
Comment (joshmh) [inline lib/blockchain-utils.js]: Do we need a whole module just for this?

**Review by joshmh — CHANGES_REQUESTED** [review]

**Pull request merged**